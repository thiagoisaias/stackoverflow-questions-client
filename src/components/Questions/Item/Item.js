import React from "react";
import PropTypes from "prop-types";
import moment from "moment";

import {
  Wrapper,
  InfoWrapper,
  BottomInfoWrapper,
  PostedBy,
  QuestionTitle,
  StatsWrapper,
  Stats,
  TagList,
  Tag,
  ProfilePictureThumbnail
} from "./styled";

const Item = props => {
  const { createdAt, numberOfAnswers, user, score, tags, title, url } = props;
  return (
    <Wrapper>
      <StatsWrapper>
        <Stats>
          <span>{score}</span>
          <span>score</span>
        </Stats>
        <Stats>
          <span>{numberOfAnswers}</span>
          <span>answers</span>
        </Stats>
      </StatsWrapper>
      <InfoWrapper>
        <a href={url} target="_blank" rel="noopener">
          <QuestionTitle>{title}</QuestionTitle>
        </a>
        <BottomInfoWrapper>
          <TagList>
            {tags && tags.map(tag => <Tag key={tag}>{tag}</Tag>)}
          </TagList>
          <PostedBy>
            <ProfilePictureThumbnail src={user && user.profileImage} />
            <span>{`Asked ${moment
              .unix(createdAt)
              .format("MMM D YYYY")} by ${user && user.name}`}</span>
          </PostedBy>
        </BottomInfoWrapper>
      </InfoWrapper>
    </Wrapper>
  );
};

Item.propTypes = {
  id: PropTypes.number,
  createdAt: PropTypes.string,
  numberOfAnswers: PropTypes.number,
  user: PropTypes.shape({
    id: PropTypes.number,
    name: PropTypes.string,
    profileImage: PropTypes.string
  }),
  score: PropTypes.number,
  tags: PropTypes.arrayOf(PropTypes.string),
  title: PropTypes.string,
  url: PropTypes.string
};

export default Item;
