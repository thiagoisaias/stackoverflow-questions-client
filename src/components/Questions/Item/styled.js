import styled from "styled-components";
import { devices } from "../../../utils/devices";

export const Wrapper = styled.div`
  display: flex;
  border-bottom: 1px solid #e2e2e2;

  @media ${devices.mediumDown} {
    padding: 16px 8px;
  }

  @media ${devices.largeUp} {
    padding: 16px;
  }
`;

export const InfoWrapper = styled.div`
  width: 100%;
  display: flex;
  flex-direction: column;
`;

export const BottomInfoWrapper = styled.div`
  display: flex;
  flex-direction: column;
`;

export const ProfilePictureThumbnail = styled.img`
  display: block;
  margin-right: 8px;

  @media ${devices.mediumDown} {
    width: 27px;
    height: 27px;
  }

  @media ${devices.largeUp} {
    width: 36px;
    height: 36px;
  }
`;

export const PostedBy = styled.span`
  display: flex;
  align-items: center;
  max-width: 250px;
  padding: 8px;
  background-color: #e1ecf4;
  color: #39739d;

  > span {
    font-size: 14px;
    color: #39739d;
  }

  @media ${devices.mediumDown} {
    margin-top: 8px;
  }
`;

export const QuestionTitle = styled.p`
  color: #484848;
  font-size: 18px;
  margin: 0;
  &:hover {
    cursor: pointer;
    opacity: 0.7;
  }
`;

export const StatsWrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;

  @media ${devices.mediumDown}{
    margin-right: 16px;
  }
  
  @media ${devices.largeUp}{
    margin-right: 48px;
  }
`;

export const Stats = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;

  &:first-child {
    margin-bottom: 16px;
    > span:first-child {
      font-size: 24px;
      color: #777;
    }

    > span:last-child {
      color: #777;
    }
  }

  &:last-child {
    background-color: #9cc670;
    color: #fff;
    padding: 12px;
    > span:first-child {
      font-size: 24px;
    }
    > span:last-child {
      font-size: 14px;
    }
  }
`;

export const TagList = styled.div`
  display: flex;
  align-items: center;
  flex-wrap: wrap;

  @media ${devices.largeUp} {
    margin: 16px 0;
  }
`;

export const Tag = styled.div`
  background-color: #e1ecf4;
  color: #39739d;
  padding: 6px;
  font-size: 13px;
  margin-right: 12px;

  @media ${devices.mediumDown} {
    padding: 6px;
    font-size: 13px;
    margin-top: 6px;
    margin-bottom: 6px;
  }

  @media ${devices.largeUp} {
    padding: 8px;
    font-size: 14px;
    margin-top: 6px;
  }
`;
