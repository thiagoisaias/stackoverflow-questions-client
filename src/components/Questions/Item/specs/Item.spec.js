import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import Item from "../Item";

describe("Item", () => {
  const mockProps = {
    id: 534,
    createdAt: "12'17",
    numberOfAnswers: 54,
    user: {
      id: 234,
      name: "Lorem",
      profileImage: "http://someurl.com"
    },
    score: 568,
    tags: ["tag1", "tag2"],
    title: "This is an example!",
    url: "http://thisisanurl.com"
  };

  const component = shallow(<Item {...mockProps} />);

  it("renders properly", () => {
    expect(toJson(component)).toMatchSnapshot();
  });
});
