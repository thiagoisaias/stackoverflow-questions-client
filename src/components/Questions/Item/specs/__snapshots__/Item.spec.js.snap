// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item renders properly 1`] = `
<styled.div>
  <styled.div>
    <styled.div>
      <span>
        568
      </span>
      <span>
        score
      </span>
    </styled.div>
    <styled.div>
      <span>
        54
      </span>
      <span>
        answers
      </span>
    </styled.div>
  </styled.div>
  <styled.div>
    <a
      href="http://thisisanurl.com"
      rel="noopener"
      target="_blank"
    >
      <styled.p>
        This is an example!
      </styled.p>
    </a>
    <styled.div>
      <styled.div>
        <styled.div
          key="tag1"
        >
          tag1
        </styled.div>
        <styled.div
          key="tag2"
        >
          tag2
        </styled.div>
      </styled.div>
      <styled.span>
        <styled.img
          src="http://someurl.com"
        />
        <span>
          Asked Invalid date by Lorem
        </span>
      </styled.span>
    </styled.div>
  </styled.div>
</styled.div>
`;
